datafusion/src/logical_plan/builder.rs

```diff
@@ -1162,6 +1164,117 @@ pub(crate) fn expand_qualified_wildcard(
     expand_wildcard(&qualifier_schema, plan)
 }
 
+/// check whether the logical plan we are building is valid
+fn check_plan_invalid(plan: &LogicalPlan) -> Result<()> {
+    match plan {
+        LogicalPlan::Projection(Projection { expr, input, .. })
+        | LogicalPlan::Sort(Sort { expr, input })
+        | LogicalPlan::Window(Window {
+            window_expr: expr,
+            input,
+            ..
+        }) => check_plan_invalid(input)
+            .and_then(|_| check_any_invalid_expr(expr, input.schema())),
+
+        LogicalPlan::Filter(Filter {
+            predicate: expr,
+            input,
+        }) => check_plan_invalid(input)
+            .and_then(|_| check_invalid_expr(expr, input.schema())),
+
+        LogicalPlan::Aggregate(Aggregate {
+            input,
+            group_expr,
+            aggr_expr,
+            ..
+        }) => {
+            let schema = input.schema();
+            check_plan_invalid(input)
+                .and_then(|_| check_any_invalid_expr(group_expr, schema))
+                .and_then(|_| check_any_invalid_expr(aggr_expr, schema))
+        }
+
+        LogicalPlan::Join(Join { left, right, .. })
+        | LogicalPlan::CrossJoin(CrossJoin { left, right, .. }) => {
+            check_plan_invalid(left).and_then(|_| check_plan_invalid(right))
+        }
+
+        LogicalPlan::Repartition(Repartition { input, .. })
+        | LogicalPlan::Limit(Limit { input, .. })
+        | LogicalPlan::Explain(Explain { plan: input, .. })
+        | LogicalPlan::Analyze(Analyze { input, .. }) => check_plan_invalid(input),
+
+        LogicalPlan::Union(Union { inputs, .. }) => {
+            inputs.iter().try_for_each(check_plan_invalid)
+        }
+
+        LogicalPlan::TableScan(TableScan {
+            table_name: _,
+            source,
+            projection,
+            projected_schema,
+            filters,
+            limit: _,
+        }) => {
+            if let Some(projection) = projection {
+                if projection.len() > projected_schema.fields().len() {
+                    return Err(DataFusionError::Plan(
+                        "projection contains columns that doesnt belong to projected schema"
+                            .to_owned(),
+                    ));
+                }
+            }
+            let schema = &source.schema().to_dfschema_ref()?;
+            check_any_invalid_expr(filters, schema)
+        }
+        _ => Ok(()),
+    }
+}
+
+/// find first error in the exprs
+fn check_any_invalid_expr(exprs: &[Expr], schema: &DFSchemaRef) -> Result<()> {
+    exprs.iter().try_for_each(|e| check_invalid_expr(e, schema))
+}
+
+/// do some checks for exprs in a logical plan
+fn check_invalid_expr(expr: &Expr, schema: &DFSchemaRef) -> Result<()> {
```

Comment:
I think I may have missed it, but as written I don't think this recurses into the `expr` -- so while it will find nonsense like

```
NOT "foo"
```

I don't think it would find
```
A = 5 OR (NOT "foo")
```

We could use an `ExpressionVisitor` for thishttps://github.com/apache/arrow-datafusion/blob/634252b44cec8dc904e48926d0aa54feeb4d48af/datafusion/src/logical_plan/expr_visitor.rs#L32-L47

Here is an example of using `ExpressionVisitor` for something similar: 
https://github.com/apache/arrow-datafusion/blob/2d6addd4c435123e934ce04564d8cc77bf101c37/datafusion/src/datasource/listing/helpers.rs#L55-L119

Reply:
Yes, this check is incomplete, it only does some simple check now.

Reply:
Yes, this function is incomplete now. If we add 
```
Expr::BinaryExpr {left, op, right} => check_type_matched(left, op, right)
```
`A = 5 OR (NOT "foo")` would be find.

I agree that visitor pattern is better, I'll try that.